File: src/cpp/study/circuit_optimization/SEValidationTarget.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include <numeric>

#include "SEValidationTarget.h"

#include "cdm/utils/GeneralMath.h"

SEValidationTarget::SEValidationTarget(Logger* logger) : Loggable(logger)
{
  m_Type = eValidationTargetType::Mean;
}

eValidationTargetType SEValidationTarget::GetType() const
{
  return m_Type;
}
void SEValidationTarget::SetType(eValidationTargetType t)
{
  m_Type = t;
}

SEDataRequest* SEValidationTarget::GetDataRequest() const
{
  return m_DataRequest;
}
void SEValidationTarget::SetDataRequest(SEDataRequest* dr)
{
  m_DataRequest = dr;
}

double SEValidationTarget::GetRangeMin() const
{
  return m_RangeMin;
}
void SEValidationTarget::SetRangeMin(double d)
{
  m_RangeMin = d;
}

double SEValidationTarget::GetRangeMax() const
{
  return m_RangeMax;
}
void SEValidationTarget::SetRangeMax(double d)
{
  m_RangeMax = d;
}

std::vector<double>& SEValidationTarget::GetData()
{
  return m_Data;
}

double SEValidationTarget::GetError() const
{
  return m_Error;
}
bool SEValidationTarget::ComputeError()
{
  m_TypeValue = SEScalar::dNaN();
  switch (m_Type)
  {
  case eValidationTargetType::Min:
  {
    m_TypeValue = *std::min_element(m_Data.begin(), m_Data.end());
    break;
  }
  case eValidationTargetType::Max:
  {
    m_TypeValue = *std::max_element(m_Data.begin(), m_Data.end());
    break;
  }
  case eValidationTargetType::Mean:
  {
    m_TypeValue = std::reduce(m_Data.begin(), m_Data.end()) / m_Data.size();
    break;
  }
  default:
  {
    Error("Unsupported validation target type!");
    return false;
  }
  }
  // NOTE: We could use PercentTolerance too
  double minError = GeneralMath::PercentDifference(m_RangeMin, m_TypeValue);
  double maxError = GeneralMath::PercentDifference(m_RangeMax, m_TypeValue);
  // No error if we are in range
  if (m_TypeValue >= m_RangeMin && m_TypeValue <= m_RangeMax)
  {
    m_Error = 0;
    return true;
  }
  // Keep the bigger error
  m_Error = (minError > maxError) ? minError : maxError;
  if (std::abs(m_Error) < 1e-15)
    m_Error = 0; // Close enough
  return true;
}

File: src/java/com/kitware/pulse/cdm/datarequests/SEValidationTarget.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package com.kitware.pulse.cdm.datarequests;

import com.kitware.pulse.cdm.bind.Engine.DataRequestData.eCategory;
import com.kitware.pulse.cdm.bind.Engine.ValidationTargetData;
import com.kitware.pulse.cdm.bind.Engine.ValidationTargetData.eType;

public class SEValidationTarget extends SEDataRequest
{
  public SEValidationTarget(eCategory c)
  {
    super(c);
  }

  public double        rangeMin;
  public double        rangeMax;
  public eType         type;
  
  public void clear()
  {
    super.clear();
    type = eType.Mean;
    rangeMin=Double.NaN;
    rangeMax=Double.NaN;
  }
  
  public static void load(ValidationTargetData src, SEValidationTarget dst)
  {
    SEDataRequest.load(src.getDataRequest(), dst);
    dst.rangeMax = src.getRangeMax();
    dst.rangeMin = src.getRangeMin();
    dst.type = src.getType();
  }
  public static ValidationTargetData unload(SEValidationTarget src)
  {
    ValidationTargetData.Builder dst = ValidationTargetData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEValidationTarget src, ValidationTargetData.Builder dst)
  {
    SEDataRequest.unload(src,dst.getDataRequestBuilder());
    dst.setRangeMax(src.rangeMax);
    dst.setRangeMin(src.rangeMin);
    dst.setType(src.type);
  }

  @Override
  public String toString()
  {
    return "Validation Target"
        + "\n\tDataRequest: " + super.toString()
        + "\n\tType: " + type
        + "\n\tRange : [" + rangeMin + ", " + rangeMax + "]";
  }
}
